odoo/addons/base/models/ir_cron.py

```diff
@@ -290,91 +300,182 @@ def _acquire_one_job(cls, cr, job_ids):
-                ir_cron, fields.Datetime.to_datetime(job['nextcall']))
-
-            # Compute how many call were missed
-            missed_call = past_nextcall
-            missed_call_count = 0
-            while missed_call <= now:
-                missed_call += interval
-                missed_call_count += 1
-            future_nextcall = missed_call
-
-            # Compute how many time we should run the cron
-            effective_call_count = (
-                     1 if not missed_call_count                    # run at least once
-                else 1 if not job['doall']                         # run once for all
-                else missed_call_count if job['numbercall'] == -1  # run them all
-                else min(missed_call_count, job['numbercall'])     # run maximum numbercall times
-            )
-            call_count_left = max(job['numbercall'] - effective_call_count, -1)
+            env = api.Environment(job_cr, job['user_id'], {'lastcall': job['lastcall'], 'cron_id': job['id']})
+            cron = env[cls._name].browse(job['id'])
+            status = None
+            for i in range(MAX_BATCH_PER_CRON_JOB):
+                cron = cron._add_progress(timed_out_counter=timed_out_counter)
+                progress = cron._get_progress()
+                job_cr.commit()
+
+                try:
+                    cron._callback(job['cron_name'], job['ir_actions_server_id'])
+                except Exception:
+                    # If some progress has been committed, we do not consider it a failure
+                    status = 'PARTIALLY_DONE' if progress.done and progress.remaining else 'FAILED'
+                else:
+                    if not progress.done or not progress.remaining:
+                        status = 'FULLY_DONE'
+                    else:
+                        status = 'PARTIALLY_DONE'
+                finally:
+                    progress.timed_out_counter = 0
+                    timed_out_counter = 0
+                    job_cr.commit()
+                _logger.info('Job `%s` (%s): Processed %s records, %s records remaining.',
+                             job['cron_name'], job['id'], progress.done, progress.remaining)
+                if status in ['FULLY_DONE', 'FAILED']:
+                    break
+
+        return status
+
+    def _update_failure_count(self, cr, job, status):
+        """ Update cron `failure_count` and `first_failure_date`.
+        Count is increased on failure and both fields are reset on success.
+        Cron that fail MIN_FAILURE_COUNT_BEFORE_DEACTIVATION times in a row over a period at least greater
+        than MIN_DELTA_BEFORE_DEACTIVATION are deactivated.
+        """
+        now = fields.Datetime.context_timestamp(self, datetime.utcnow())
+
+        if status == 'FAILED':
+            failure_count = job['failure_count'] + 1
+            first_failure_date = job['first_failure_date'] or now
+            active = True
```

Comment:
I have a bad feeling about forcing `active=True`, I would rather we leave the original value. I fully agree that at this moment it always enters with `job['active']==True` but I would prefer to be safe than sorry...
```suggestion
            active = job['active']
```